sklearn/utils/tests/test_validation.py

```diff
@@ -1445,3 +1446,31 @@ def test_check_array_deprecated_matrix():
     )
     with pytest.warns(FutureWarning, match=msg):
         check_array(X)
+
+
+@pytest.mark.parametrize("names", [[f"col_{i}" for i in range(3)], list(range(3))])
+def test_get_feature_names_pandas(names):
+    """Get feature names with pandas dataframes."""
+    pd = pytest.importorskip("pandas")
+    X = pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns=names)
+
+    names = _get_feature_names(X)
+    assert_array_equal(names, names)
+
+
```

Comment:
BTW, do we really want to accept integer column names in dataframes or should we strict from the start and raise a `ValueError`?

Or is the plan to only force the conversion to `str` names later in a `get_features_names_in_()` / `get_features_names_out_()` method in a later PR?

Reply:
We can delay this discussion to later if we do not reach a consensus. As of now we accepts such dataframe without raising `ValueError` so if we decide to raise `ValueError` we would need a deprecation cycle anyways.

Reply:
Just saw this was also discussed in https://github.com/scikit-learn/scikit-learn/pull/18010#discussion_r667213355.

Reply:
IMO I would keep the same semantic that we have in `ColumnTransformer`: integers are used for positioning and I would not allow using them as column names. I find it really ambiguous.